client-java/src/main/java/uk/gov/gchq/palisade/client/resource/ResourceClientListener.java

```diff
@@ -0,0 +1,237 @@
+        var error = getObjectMapper().convertValue(body, Error.class);
+        post(ErrorEvent.of(error));
+    }
+
+    private void handleComplete() {
+        LOGGER.debug("handle Complete for token {}", getToken());
+        post(ResourcesExhaustedEvent.of(getToken()));
+    }
+
+    private void handleResource(final Message message) {
+        LOGGER.debug("handle resource for token {}, message: {}", getToken(), message);
+        var body = message.getBody().orElseThrow(() -> new MissingMessageBodyException(message));
+        var resource = getObjectMapper().convertValue(body, Resource.class);
+        post(ResourceReadyEvent.of(resource));
+    }
+
+    @Override
+    public void onError(final WebSocket ws, final Throwable error) {
+        LOGGER.debug("A {} exception was thrown.", error.getCause().getClass().getSimpleName());
+        LOGGER.debug("Message: {}", error.getLocalizedMessage());
+    }
+
+    @Override
+    public void onOpen(final WebSocket ws) {
+        Listener.super.onOpen(ws);
+        LOGGER.debug("WebSocket Listener has been opened for requests.");
+        sendMessage(ws, b -> b.type(MessageType.CTS));
+    }
+
+    @Override
+    public CompletionStage<?> onText(final WebSocket ws, final CharSequence data, final boolean last) {
+
+        var text = data.toString();
+
+        ws.request(1); // omit this and no methods are called listener
+
+        Message message;
+        try {
+            message = getObjectMapper().readValue(text, Message.class);
+        } catch (JsonProcessingException e) {
+            onError(ws, e);
+            return null;
+        }
+
+        LOGGER.debug("Recd: {}", message);
+
+        var type = message.getType();
+
+        if (type == MessageType.RESOURCE) {
+            handleResource(message);
+        } else if (type == MessageType.COMPLETE) {
+            handleComplete();
+        } else if (type == MessageType.ERROR) {
+            handleError(message);
+        } else {
+            // ignore the unsupported message
+            LOGGER.warn("Ignoring unsupported {} message type", type);
+        }
+
+        if (type != MessageType.COMPLETE) {
+
+            LOGGER.debug("Testing if CTS can be sent for token {}", getToken());
+
+            // This is a quite crude way of waiting for download slots to become available
+            // Should implement a better way, but this will do for now.
+
+            while (!getDownloadTracker().hasAvailableSlots()) {
+                try {
+                    LOGGER.debug("No download slots available, waiting");
+                    Thread.sleep(ONE_SECOND);
+                } catch (InterruptedException e) { // just swallow this
+                    Thread.currentThread().interrupt();
+                    LOGGER.warn("This thread was sleeping, when it was interrupted: {}", e.getMessage());
+                    // we'll loop round to see if there are any available slots
+                }
```

Comment:
The problem I was having here is that the gap between resources could be quite large. I'm not really sure how long to wait. I think this may need to be handled better. Maybe when a download completes, if anothere download can be scheduled, then a CTS is sent. This would mean there are no waits. 

I'll look into this as it would be a much better solution.

Reply:
Fixed